cloudinit/net/network_manager.py

```diff
@@ -189,40 +195,38 @@ def _add_address(self, family, subnet):
         value = subnet["address"] + "/" + str(subnet["prefix"])
         self._add_numbered(family, "address", value)
 
-    def _add_route(self, family, route):
-        """
-        Adds a ipv[46].route<n> property.
-        """
-
+    def _add_route(self, route):
+        """Adds a ipv[46].route<n> property."""
+        # Because network v2 route definitions can have mixed v4 and v6
+        # routes, determine the family per route based on the gateway
+        family = "ipv6" if is_ipv6_network(route["gateway"]) else "ipv4"
         value = route["network"] + "/" + str(route["prefix"])
         if "gateway" in route:
             value = value + "," + route["gateway"]
         self._add_numbered(family, "route", value)
 
-    def _add_nameserver(self, dns):
+    def _add_nameserver(self, dns: str) -> None:
         """
         Extends the ipv[46].dns property with a name server.
         """
 
-        # FIXME: the subnet contains IPv4 and IPv6 name server mixed
-        # together. We might be getting an IPv6 name server while
-        # we're dealing with an IPv4 subnet. Sort this out by figuring
-        # out the correct family and making sure a valid section exist.
         family = "ipv6" if is_ipv6_address(dns) else "ipv4"
         self._set_default(family, "method", "disabled")
-
         self._set_default(family, "dns", "")
         self.config[family]["dns"] = self.config[family]["dns"] + dns + ";"
 
-    def _add_dns_search(self, family, dns_search):
+    def _add_dns_search(self, dns_search: List[str]) -> None:
         """
         Extends the ipv[46].dns-search property with a name server.
         """
-
-        self._set_default(family, "dns-search", "")
-        self.config[family]["dns-search"] = (
-            self.config[family]["dns-search"] + ";".join(dns_search) + ";"
-        )
+        for family in ["ipv4", "ipv6"]:
+            if self.config.has_section(family):
```

Comment:
This means that a configuration that specifies no addresses will now have no search domains rendered, correct?

I'm curious what netplan does in this case, since it seems like an invalid configuration, but I think that this behavior makes sense.

Reply:
```
root@me:/etc/netplan# cat 50-cloud-init.yaml 
network:
  version: 2
  renderer: NetworkManager
  ethernets:
    eth0:
      nameservers:
        search: [lab, home]
        addresses: [8.8.8.8, FEDC::1]
root@me:/etc/netplan# cat /run/NetworkManager/system-connections/netplan-eth0.nmconnection
[connection]
id=netplan-eth0
type=ethernet
interface-name=eth0

[ethernet]
wake-on-lan=0

[ipv4]
method=link-local
dns=8.8.8.8;

[ipv6]
method=manual
ip6-privacy=0
dns=FEDC::1;
dns-search=lab;home;
```

I have yet to make any conclusions about it :smile:

Reply:
> I have yet to make any conclusions about it 😄

Errr, why isn't netplan adding `dns-search` to the ipv4 entry? Is that a netplan bug?

Reply:
Yeah, I'm not sure about that either. I'll file a bug and see. 

I did push an update though. This one waits until we've processed all of the subnets, and then adds the dns search entries. That means if we haven't defined ipv6 (or ipv4) at all, then we won't try to add the dns-search entry to that family.

Reply:
> I'll file a bug

Jk, found this: https://bugs.launchpad.net/netplan/+bug/1786726

Reply:
Gotcha, that explains the behavior thanks